Jamboree dimensionality reduction methods

lazappi opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Submission type

* [ ] This submission adds a new dataset
* [x] This submission adds a new method
* [ ] This submission adds a new metric
* [ ] This submission adds a new task
* [ ] This submission adds a new Docker image
* [ ] This submission fixes a bug (link to related issue: )
* [ ] This submission adds a new feature not listed above

### Testing

* [x] This submission was written on a forked copy of SingleCellOpenProblems
* [x] GitHub Actions "Run Benchmark" tests are passing on this base branch of this pull request (include link to passed test: )
* [x] If this pull request is not ready for review (including passing the "Run Benchmark" tests), I will open this PR as a draft (click on the down arrow next to the "Create Pull Request" button)

### Submission guidelines

* [x] This submission follows the guidelines in our [Contributing](../blob/master/CONTRIBUTING.md) document
* [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change

### PR review checklist

This PR will be evaluated on the basis of the following checks:

* [x] The task addresses a valid open problem in single-cell analysis
* [x] The latest version of master is merged and tested
* [x] The methods/metrics are imported to `__init__.py` and were tested in the pipeline
* [x] Method and metric decorators are annotated with paper title, year, author, code version, and date
* [x] The README gives an outline of the methods, metrics and datasets in the folder
* [ ] The README provides a satisfactory task explanation (for new tasks)
* [ ] The sample test data is appropriate to test implementation of all methods and metrics (for new tasks)

# Description

This PR contains the following new dimensionality reduction methods from the jamboree:

<table>
<tr>
	<td><strong>Method</strong></td>
	<td><strong>Author</strong></td>
	<td><strong>Issue</strong></td>
	<td><strong>Original PR</strong></td>
</tr>
<tr>
	<td>DensMAP</td>
	<td>@bendemeo</td>
	<td>#249</td>
	<td>lazappi/SingleCellOpenProblems/pull/3</td>
</tr>
<tr>
	<td>NeuralEE</td>
	<td>@michalk8</td>
	<td>#284</td>
	<td>lazappi/SingleCellOpenProblems/pull/4</td>
</tr>
<tr>
	<td>IVIS</td>
	<td>@michalk8</td>
	<td>#270</td>
	<td>lazappi/SingleCellOpenProblems/pull/9</td>
</tr>
<tr>
	<td>scvis</td>
	<td>@michalk8</td>
	<td>#292</td>
	<td>lazappi/SingleCellOpenProblems/pull/5</td>
</tr>
</table>

mvinyard reviewed (approved):
This looks great to me @lazappi, thanks!

LuckyMD reviewed (commented):
I wonder if `preprocess_scanpy` should be a part of the data loader rather than the method here. You seem to be relying on all methods doing the same preprocessing to fairly compare. If you don't do that, you could instead call the methods for example: `alra_logCPM_1kHVG`. Although I'd prefer the data-loader side i think

lazappi commented:
I vaguely remember this being discussed at some point but I can't remember why we did it this way... Maybe there is more consistency with how this is being handled across the different tasks now though?

LuckyMD commented:
We do deal with the preprocessing in the methods this way in batch integration as well. But then we have naming conventions for the methods that reflect this. So it would be great if it could either be put into the data loader or into the name. Would you still have scope to adapt this @lazappi ?

lazappi commented:
In theory probably but depends how quickly you want it done (and how quickly I can member how this works 😸)

LuckyMD commented:
We're trying to get most of the PRs merged by the end of the month ideally.

lazappi commented:
Ok, I think I have done this now.

Changes:

- Rename `preprocess_scanpy()` to `preprocess_logCPM_1kHVG()` for clarity
- Modified `preprocess_logCM_1kHVG()` to subset to HVGs (I think this is the only way to be sure that methods are using only these genes but can undo if needed)
- Renamed methods to include the preprocessing as a suffix
- Added two new method versions I thought were missing (`neuralee_logCPM_1kHVG` (current version had the methods suggested default processing) and `densmap_pca_logCPM_1kHVG` (current version was run on expression not PCA input)). Hopefully they pass the checks 🤞🏻.

lazappi commented:
Any suggestions on why the tests failed? I have forgotten most of how this goes together. Maybe I need to merge an update from somewhere into here?

LuckyMD commented:
Hey @lazappi,
I also noticed that the GA testing workflow needs to be fixed. @scottgigante will be looking into fixing this in the next couple of weeks. When it's fixed on master you can merge back into this PR and hopefully we can get it merged.

LuckyMD commented:
@lazappi the testing workflow should be fixed now with #324

lazappi commented:
> 1. Should we move this method to `normalize` rather than having it just inside this task? Seems more generally useful.

That makes sense to me but I guess it's up to the core people how much you want to standardise things across tasks. I think we probably avoided that during the Jamboree because we didn't want to touch parts of the code that other people were using.

> 2. Should we change all normalizations to return the anndata to allow for normalizations that are not in-place?

I don't really have an opinion on this, not sure whether it is better to be consistent or do things in-place where possible (maybe should be a separate PR though?).

scottgigante-immunai commented on docker/openproblems-python-method-scvis/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+FROM python:3.6
```
Should we create an `openproblems-py36` image as a base for this? Seems broader than just scvis.

lazappi replied:
That seems reasonable to me I suppose. Is it just renaming things or would there need to be bigger updates?

scottgigante-immunai replied:
The question would be whether we follow the structure in the py38 image in which we have a python base container and then an `extras` container built on top that adds stuff like scvis.

LuckyMD replied:
For 3.6 I reckon it's probably sth that will not be heavily used anymore so a separate image should be enough, no? How much python 3.6 dependencies are flying about?

scottgigante replied:
That seems reasonable. Let's just call it openproblems-python36 and leave it at that for now

lazappi replied:
Ok, I renamed this to `openproblems-python36`

scottgigante-immunai commented on openproblems/tasks/dimensionality_reduction/methods/densmap.py:
```diff
@@ -0,0 +1,41 @@
+    code_version=check_version("umap-learn"),
+    image="openproblems-python-extras",
+)
+def densmap_logCPM_1kHVG(adata):
```
per #330 these methods will all have to have a kwarg `test: bool=False`

lazappi replied:
Added the argument to all the methods but not currently used.

scottgigante-immunai replied:
No worries, I don't think it's necessary to use it in all methods

scottgigante-immunai commented on openproblems/tasks/dimensionality_reduction/methods/preprocessing.py:
```diff
@@ -1,11 +1,15 @@
 
 
-def preprocess_scanpy(adata):
+def preprocess_logCPM_1kHVG(adata: AnnData) -> AnnData:
```
Move to `openproblems.tools.normalize` per #329

lazappi replied:
Done, and (hopefully) updated in all the methods.

lazappi commented on docker/openproblems-python-method-scvis/README.md:
```diff
@@ -0,0 +1,18 @@
+# openproblems Docker image
```
Changed the name in this README

lazappi commented:
I think the errors in the checks are coming from the `normalizer` decorator. Maybe I wasn't supposed to add this? Any other hints about the best way to fix this?

scottgigante-immunai commented:
Hmm, this is odd. Seems like a problem with scanpy's HVG code? By moving this to the normalizers submodule you added it to the relevant unit tests, which are what's broken. However the tests don't yet reflect the fact that this returns an anndata because #329 hasn't been merged yet, so perhaps once that merge happens this might resolve itself.

> disp_cut_off = dispersion_norm[n_top_genes - 1]
> E           IndexError: index 29 is out of bounds for axis 0 with size 21
> 
> /opt/hostedtoolcache/Python/3.8.12/x64/lib/python3.8/site-packages/scanpy/preprocessing/_highly_variable_genes.py:268: IndexError

scottgigante-immunai commented:
Update: #329 is now merged, pull from `main` and let's see if we have the same issue.

lazappi commented:
Ok, so checks are failing now because one of the metrics that was already merged into `main` (and has now been merged here) used the old pre-processing function. I can replace it with the new one which would fix the error but I think there is a bigger problem.

Methods have likely processed the data in some way but there is no way of knowing how when calculating metrics. How do we make sure that metrics which need the original data as well as whatever the task output is (in this case the embedding) are provided with that data in a standard format? Resetting the `X` matrix to the original version before returning from the method would probably work here but not sure that would work for every task. Storing the original `X` as a layer would maybe be a better option (maybe even as part of the data loader).

Thoughts?

(Also let me know if that isn't clear or I have misunderstood something)

scottgigante-immunai commented:
This makes sense. I have a vague memory of `adata.layers['counts']` being used for this but I don't see anywhere where we actually populate this. I would propose we add a `adata.layers['counts'] = adata.X` immediately after loading the data in https://github.com/openproblems-bio/openproblems/blob/main/openproblems/data/utils.py#L49. That way the raw data is always accessible. Thoughts?

lazappi commented:
That sounds mostly reasonable to me but I'm not sure how it would affect the other tasks. Does everyone assume that `adata.X` contains counts when the data is first loaded? I also wonder how obvious this would be to any future contributors. How would they know this happens whenever a dataset is loaded?

scottgigante-immunai commented:
Certainly we assume that `adata.X` is not normalized (see [CONTRIBUTING](https://github.com/openproblems-bio/openproblems/blob/main/CONTRIBUTING.md#adding-a-dataset--method--metric-to-a-task)) -- we could instead call it `adata.layers["raw"]` if we wanted to, but I think "counts" is pretty standard.

scottgigante-immunai commented on openproblems/tools/normalize.py:
```diff
@@ -52,3 +52,15 @@ def sqrt_cpm(adata: ad.AnnData) -> ad.AnnData:
+
+    log_cpm(adata)
+    sc.pp.highly_variable_genes(adata, n_top_genes=1000, flavor="cell_ranger")
+    adata = adata[:, adata.var["highly_variable"]]
```
Shouldn't we call `.copy()` here since otherwise we're just creating a view?

scottgigante-immunai commented on openproblems/tools/normalize.py:
```diff
@@ -52,3 +52,15 @@ def sqrt_cpm(adata: ad.AnnData) -> ad.AnnData:
+    sc.pp.highly_variable_genes(adata, n_top_genes=1000, flavor="cell_ranger")
+    adata = adata[:, adata.var["highly_variable"]]
+    sc.tl.pca(adata, n_comps=50, svd_solver="arpack")
+    adata.obsm["X_input"] = adata.obsm["X_pca"]
```
I don't think this actually makes sense inside a normalizer -- PCA is a preprocessing step but not a normalization step (the normalizer should only touch `adata.X`). Can we move PCA back inside the methods rather than inside the normalizer?

Most obvious problem here is the use of `preprocess_logCPM_1kHVG(adata, obsm="mode2")` which would compute the PCA but not store it anywhere due to the way we handle normalization of `obsm` which is by creating a new AnnData, normalizing `.X`, and pasting the normalized `.X` back onto `obsm` of the original.

lazappi replied:
I vaguely remember this coming from Daniel but yes, happy to move it.

scottgigante commented:
@lazappi https://github.com/openproblems-bio/openproblems/pull/343 is now merged, this should solve your problem yeah?

scottgigante-immunai commented on docker/openproblems-python36/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+RUN ./aws/install
+
+# install dependencies
+COPY ./docker/openproblems-python-method-scvis/requirements.txt ./requirements.txt
```
This should be `openproblems-python36`

lazappi commented:
Any suggestions on the failing checks? I tried locally but wasn't able to reproduce it.

LuckyMD commented:
Just to add to the discussion above that I'm only seeing now: you can define the task data expectations in your README. You should define you task API there, which includes specifying the fields you expect task dataset functions to have when new datasets are added. See here for an example: https://github.com/openproblems-bio/openproblems/blob/575b9979d8764c58bb7ab2ffb70fd2aba787961b/openproblems/tasks/_batch_integration/batch_integration_graph/README.md

scottgigante-immunai commented:
@lazappi maybe related: https://github.com/scverse/scanpy/issues/834. I'd suggest you put in some print statements on `adata.shape` and `n_var_genes` inside the normalizer to debug inside the test suite.

also note `neuralee_default` needs a `test` kwarg: https://github.com/openproblems-bio/openproblems/runs/5829492305?check_suite_focus=true#step:14:1359

lazappi commented:
> Just to add to the discussion above that I'm only seeing now: you can define the task data expectations in your README. You should define you task API there, which includes specifying the fields you expect task dataset functions to have when new datasets are added. See here for an example: [`575b997`/openproblems/tasks/_batch_integration/batch_integration_graph/README.md](https://github.com/openproblems-bio/openproblems/blob/575b9979d8764c58bb7ab2ffb70fd2aba787961b/openproblems/tasks/_batch_integration/batch_integration_graph/README.md)

Yeah, now that I look at it the README could do with an update. I must have forgotten to make people do that as part of their individual pull requests. Do you think we still need to describe the `adata.layers['counts']` thing here given that they way it is implemented it will happen for all tasks (or better just to do it anyway for clarity)?

LuckyMD commented:
> Do you think we still need to describe the adata.layers['counts'] thing here given that they way it is implemented it will happen for all tasks (or better just to do it anyway for clarity)?

I would detail all expected slots in the README, as this slot could otherwise always be removed by someone in the task dataset function. If you detail it in the README, you should also add a test for it in the task.

lazappi commented:
I think the only test that is failing now is because of SCOT from the multimodal task

scottgigante-immunai reviewed (changes requested):
All LGTM, except for this comment:

>  GitHub Actions "Run Benchmark" tests are passing on this base branch of this pull request (include link to passed test: )

No link to passed test and as far as I can see it's not passing. Try merging `main` and see if that solves the SCOT issue.

lazappi commented:
Merging `main` doesn't seem to have helped the SCOT error, any other suggestions? **ivis** is failing the benchmarks on my fork so I'll try to sort that out.

scottgigante-immunai commented:
Currently working on a [PR](https://github.com/openproblems-bio/openproblems/pull/364) to fix this [issue](https://github.com/lazappi/openproblems/runs/6080675531?check_suite_focus=true#step:12:34) with codecov. Stems from the fact that you haven't authorized codecov for your fork, which we shouldn't care about.

scottgigante-immunai commented on openproblems/tools/normalize.py:
```diff
@@ -52,3 +53,19 @@ def sqrt_cpm(adata: ad.AnnData) -> ad.AnnData:
+
+
+@decorators.normalizer
+def preprocess_logCPM_1kHVG(adata: ad.AnnData) -> ad.AnnData:
```
Can we refactor this to `log_cpm_hvg(adata: ad.AnnData, n_genes: int = 1000)` to a) be consistent with the naming style in the rest of the file and b) allow users to change n_genes?

scottgigante-immunai commented on openproblems/tasks/dimensionality_reduction/methods/tsne.py:
```diff
@@ -15,8 +15,9 @@
     code_version=check_version("MulticoreTSNE"),
     image="openproblems-python-extras",
 )
-def tsne(adata, test=False, n_pca=50):
```
Question: why did we remove these methods rather than adding HVG as an additional method?

lazappi replied:
I just renamed them when we changed the preprocessing function. Happy to have other variants, just need to decide what preprocessing they should have.

scottgigante-immunai replied:
I think it's fine as is.

scottgigante-immunai reviewed (changes requested):
We're very close! Just a couple of nitpicks.

lazappi commented:
> Merging `main` doesn't seem to have helped the SCOT error, any other suggestions? **ivis** is failing the benchmarks on my fork so I'll try to sort that out.

Any suggestions for working out what is happening with the benchmark? I tried running **ivis** in Docker locally at it ran without any issues. The latest benchmark run on my fork is here https://github.com/lazappi/openproblems/actions/runs/2182678993

scottgigante-immunai commented:
`132 Illegal instruction` is a C++ error. Probably a tensorflow problem? I notice that `ivis` requirements call for `tensorflow>=1.13.1` and we're using `tensorflow==2.4.1` -- could that be the problem?

In any case, the easiest path forward is probably to go ahead and split ivis into its own PR. That way we can merge this one and deal with ivis on its own.

lazappi commented:
Ah, I wasn't able to find the error message but that does look like it could be a **tensorflow** thing. I have moved `ivis` to another (draft) PR. Is there a way to trigger the benchmark action on my fork? It hasn't run for the updates on this https://github.com/lazappi/openproblems/actions.
